src/liballoc/collections/btree/map.rs

```diff
@@ -201,16 +201,40 @@ impl<K: Clone, V: Clone> Clone for BTreeMap<K, V> {
         }
 
         if self.is_empty() {
-            // Ideally we'd call `BTreeMap::new` here, but that has the `K:
-            // Ord` constraint, which this method lacks.
-            BTreeMap {
-                root: node::Root::shared_empty_root(),
-                length: 0,
-            }
+            BTreeMap::new()
         } else {
             clone_subtree(self.root.as_ref())
         }
     }
+
+    fn clone_from(&mut self, other: &Self) {
```

Comment:
I think this algorithm needs some comments so that you don't have to sit around and think about it for some time to understand how and why it works...